### PR TITLE
[master < T981] Reduce all shortest paths memory usage

### PR DESCRIPTION
**Pull request by antoniofilipovic**
TLDR: Based on the `mgbench` results, memory usage dropped from gigabytes to a few megabytes, while performance on all shortest paths dropped ~7% -> the decision is to merge because of the significant gain in memory usage while keeping performance high.

[master < Task] PR
- [x] Check, and update documentation if necessary
- [x] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [x] Write a release note here
- [x] Tag someone from docs team in the comments

### CODE REVIEW

**Comment by gitbuda** [comment]
@antoniofilipovic can we merge this to `master` before v2.9 is released (include in the v2.9 release)?

**Comment by Josipmrden** [comment]
please provide text for the PR message

File: src/query/cypher_query_interpreter.cpp [hunk 1]
```diff
@@ -84,7 +84,6 @@ ParsedQuery ParseQuery(const std::string &query_string, const std::map<std::stri
     if (visitor.GetQueryInfo().has_load_csv && !query_config.allow_load_csv) {
       throw utils::BasicException("Load CSV not allowed on this instance because it was disabled by a config.");
     }
-
     if (visitor.GetQueryInfo().is_cacheable) {
```
Comment (Josipmrden) [inline src/query/cypher_query_interpreter.cpp:87]: delete this change, does not impact the PR

File: src/query/interpret/eval.hpp [hunk 2]
```diff
@@ -793,15 +793,15 @@ class ExpressionEvaluator : public ExpressionVisitor<TypedValue> {
...
       if (element.IsNull()) {
         result.emplace_back();
       } else {
-        frame_->at(element_symbol) = element;
+        frame_->at(element_symbol) = std::move(element);
         result.emplace_back(extract.expression_->Accept(*this));
       }
     }
-    return TypedValue(result, ctx_->memory);
+    return TypedValue(std::move(result), ctx_->memory);
```
Comment (Josipmrden) [inline src/query/interpret/eval.hpp:804]: why is this impacting AllShortestPath if it is Extract?

File: src/query/interpreter.cpp [hunk 3]
```diff
@@ -165,6 +165,27 @@ std::optional<std::string> GetOptionalStringValue(query::Expression *expression,
...
 };
 
+bool IsAllShortestPathsQuery(const std::vector<memgraph::query::Clause *> &clauses) {
+  for (const auto &clause : clauses) {
+    if (clause->GetTypeInfo() != Match::kType) {
+      continue;
+    }
+    auto *match_clause = utils::Downcast<Match>(clause);
+    for (auto &pattern : match_clause->patterns_) {
+      for (auto &atom : pattern->atoms_) {
```
Comment (Josipmrden) [inline src/query/interpreter.cpp:175]:
```suggestion
    for (const auto &pattern : match_clause->patterns_) {
      for (const auto &atom : pattern->atoms_) {
```

File: src/query/interpreter.cpp [hunk 4]
```diff
@@ -1380,8 +1401,14 @@ PreparedQuery PrepareCypherQuery(ParsedQuery parsed_query, std::map<std::string,
...
     contains_csv = true;
   }
+
   // If this is LOAD CSV query, use PoolResource without MonotonicMemoryResource as we want to reuse allocated memory
-  auto use_monotonic_memory = !contains_csv && !IsCallBatchedProcedureQuery(clauses);
+  auto use_monotonic_memory =
+      !contains_csv && !IsCallBatchedProcedureQuery(clauses) && !IsAllShortestPathsQuery(clauses);
+
+  spdlog::trace("Has all shortest PrepareCypher {} and use monotonic {}", IsAllShortestPathsQuery(clauses),
+                use_monotonic_memory);
```
Comment (Josipmrden) [inline src/query/interpreter.cpp:1410]:
```suggestion
  spdlog::trace("PrepareCypher has {} encountered all shortest paths and {} use of monotonic memory {}", IsAllShortestPathsQuery(clauses) ? "", "not",
                use_monotonic_memory ? "", "no");
```

File: src/query/interpreter.cpp [hunk 5]
```diff
@@ -3136,14 +3167,15 @@ Interpreter::PrepareResult Interpreter::Prepare(const std::string &query_string,
...
-          })) {
+          IsAllShortestPathsQuery(clauses) || IsCallBatchedProcedureQuery(clauses) ||
+          std::any_of(clauses.begin(), clauses.end(),
+                      [](const auto *clause) { return clause->GetTypeInfo() == LoadCsv::kType; })) {
         // Using PoolResource without MonotonicMemoryResouce for LOAD CSV reduces memory usage.
         // QueryExecution MemoryResource is mostly used for allocations done on Frame and storing `row`s
         query_executions_[query_executions_.size() - 1] = std::make_unique<QueryExecution>(utils::PoolResource(
             128, kExecutionPoolMaxBlockSize, utils::NewDeleteResource(), utils::NewDeleteResource()));
         query_execution_ptr = &query_executions_.back();
+        spdlog::trace("Has all shortest QueryExecutions Resource {}", IsAllShortestPathsQuery(clauses));
```
Comment (Josipmrden) [inline src/query/interpreter.cpp:3178]: Also refactor this one in similar way to the above one with ternary operator to form a proper sentence

**Comment by Josipmrden** [comment]
small comments, overall GJ

File: src/query/interpret/eval.hpp (hunk 2 above)
Comment (antoniofilipovic) [reply src/query/interpret/eval.hpp:804]: I forgot, but it made an impact on some analysis if I am not wrong. I think I can delete, if you think I should let me know

Comment (Josipmrden) [reply src/query/interpret/eval.hpp:804]: it's fine

**Review by Josipmrden — APPROVED** [review]

**Comment by antoniofilipovic** [comment]
@vpavicic This PR changes the way memory is used in all shortest-path queries. From now on users should see lower memory usage with query performance almost staying the same.

**Pull request merged**